File: src/main/java/com/github/kindrat/cassandra/client/ui/window/menu/file/ConnectionManager.java
package com.github.kindrat.cassandra.client.ui.window.menu.file;

import com.datastax.driver.core.TypeCodec;
import com.github.kindrat.cassandra.client.i18n.MessageByLocaleService;
import com.github.kindrat.cassandra.client.properties.StorageProperties;
import com.github.kindrat.cassandra.client.properties.UIProperties;
import com.github.kindrat.cassandra.client.ui.ConnectionData;
import com.github.kindrat.cassandra.client.ui.MainController;
import com.github.kindrat.cassandra.client.ui.eventhandler.TableClickEvent;
import com.github.kindrat.cassandra.client.ui.window.menu.ConnectionDataHandler;
import com.google.common.io.Files;
import javafx.beans.property.SimpleObjectProperty;
import javafx.collections.FXCollections;
import javafx.geometry.Pos;
import javafx.scene.Scene;
import javafx.scene.control.*;
import javafx.scene.layout.VBox;
import javafx.stage.Modality;
import javafx.stage.Stage;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.StringUtils;
import org.springframework.beans.BeansException;
import org.springframework.beans.factory.BeanFactory;
import org.springframework.beans.factory.BeanFactoryAware;

import java.io.File;
import java.util.List;
import java.util.function.Consumer;
import java.util.function.Function;

import static com.github.kindrat.cassandra.client.util.JsonUtil.fromJson;
import static com.github.kindrat.cassandra.client.util.JsonUtil.toJson;
import static com.github.kindrat.cassandra.client.util.UIUtil.cellFactory;
import static java.nio.charset.StandardCharsets.UTF_8;

@Slf4j
public class ConnectionManager extends Stage implements BeanFactoryAware {
    private final MessageByLocaleService localeService;
    private final UIProperties uiProperties;
    private final StorageProperties storageProperties;
    private final TableView<ConnectionData> table;
    private BeanFactory beanFactory;
    private final File configurationsFile;

    @SneakyThrows
    public ConnectionManager(Stage parent, MessageByLocaleService localeService, UIProperties uiProperties,
            StorageProperties storageProperties) {
        this.localeService = localeService;
        this.uiProperties = uiProperties;
        this.storageProperties = storageProperties;

        initModality(Modality.APPLICATION_MODAL);
        initOwner(parent);
        setTitle(localeService.getMessage("ui.menu.file.manager.title"));
        Integer width = uiProperties.getConnectionManagerWidth();
        Integer height = uiProperties.getConnectionManagerHeight();
        Integer rightWidth = uiProperties.getConnectionManagerRightPaneWidth();
        Integer leftWidth = width - rightWidth;

        VBox connectionBox = new VBox(uiProperties.getNewConnectSpacing());
        connectionBox.setAlignment(Pos.CENTER);
        connectionBox.setMinWidth(leftWidth);
        connectionBox.setMaxWidth(leftWidth);

        table = buildConnectionDataTable(leftWidth);
        configurationsFile = getConfigurationsFile();
        if (!configurationsFile.exists()) {
            configurationsFile.createNewFile();
        }
        String json = Files.toString(configurationsFile, UTF_8);
        if (!StringUtils.isBlank(json)) {
            List<ConnectionData> connections = fromJson(json, ConnectionData.class);
            table.getItems().addAll(connections);
        }
        connectionBox.getChildren().add(table);

        VBox buttons = new VBox();
        buttons.setMinWidth(rightWidth);
        buttons.setMaxWidth(rightWidth);
        buttons.setSpacing(uiProperties.getConnectionManagerButtonsSpacing());
        buttons.setAlignment(Pos.CENTER);

        Button newConnection = getButton("+");
        newConnection.setOnAction(event -> onNewConnection());
        Button removeConnection = getButton("-");
        removeConnection.setOnAction(event -> onConnectionRemove());
        Button editConnection = getButton("✎");
        editConnection.setOnAction(event -> onConnectionEdit());
        Button useConnection = getButton("✓");
        useConnection.setOnAction(event -> onConnectionUsage());

        buttons.getChildren().add(newConnection);
        buttons.getChildren().add(removeConnection);
		buttons.getChildren().add(editConnection);
        buttons.getChildren().add(useConnection);

        SplitPane splitPane = new SplitPane(connectionBox, buttons);
        Scene content = new Scene(splitPane, width, height);

        setScene(content);
        show();
    }

    @SneakyThrows
    private File getConfigurationsFile() {
        return new File(storageProperties.getLocation(), storageProperties.getPropertiesFile());
    }

    private Button getButton(String text) {
        Button button = new Button(text);
        button.setMaxWidth(uiProperties.getConnectionManagerButtonWidth());
        button.setMinWidth(uiProperties.getConnectionManagerButtonWidth());
        return button;
    }

    private TableView<ConnectionData> buildConnectionDataTable(Integer leftWidth) {
        TableView<ConnectionData> dataTableView = new TableView<>();
        dataTableView.setMinWidth(leftWidth);
        dataTableView.setMaxWidth(leftWidth);
        dataTableView.setColumnResizePolicy(TableView.UNCONSTRAINED_RESIZE_POLICY);
        dataTableView.getSelectionModel().setSelectionMode(SelectionMode.SINGLE);
        dataTableView.getSelectionModel().setCellSelectionEnabled(true);
        dataTableView.setOnMouseClicked(new TableClickEvent<>(dataTableView));

        int colWidth = leftWidth / 4;

        String urlTitle = localeService.getMessage("ui.menu.file.manager.table.url");
        TableColumn<ConnectionData, String> url = buildColumn(urlTitle, colWidth, ConnectionData::getUrl);

        String spaceTitle = localeService.getMessage("ui.menu.file.manager.table.keyspace");
        TableColumn<ConnectionData, String> keyspace = buildColumn(spaceTitle, colWidth, ConnectionData::getKeyspace);

        String userTitle = localeService.getMessage("ui.menu.file.manager.table.username");
        TableColumn<ConnectionData, String> username = buildColumn(userTitle, colWidth, ConnectionData::getUsername);

        String passTitle = localeService.getMessage("ui.menu.file.manager.table.password");
        TableColumn<ConnectionData, String> password = buildColumn(passTitle, colWidth, ConnectionData::getPassword);

        dataTableView.getColumns().add(url);
        dataTableView.getColumns().add(keyspace);
        dataTableView.getColumns().add(username);
        dataTableView.getColumns().add(password);
        dataTableView.setItems(FXCollections.observableArrayList());

        return dataTableView;
    }

    private TableColumn<ConnectionData, String> buildColumn(String name, int columnWidth, Function<ConnectionData,
            String> extractor) {
        TableColumn<ConnectionData, String> column = new TableColumn<>();
        Label columnLabel = new Label(name);
        column.setCellFactory(cellFactory(TypeCodec.varchar()));

        column.setCellValueFactory(param -> {
            String field = extractor.apply(param.getValue());
            return new SimpleObjectProperty<>(field);
        });

        column.setOnEditCommit(event -> log.info("edit"));
        column.setGraphic(columnLabel);
        column.setMinWidth(columnWidth);
        return column;
    }

    private void onNewConnection() {
        beanFactory.getBean("newConnectionBox", this, (ConnectionDataHandler) this::onConnectionData);
    }

    private void onConnectionData(ConnectionData data) {
        table.getItems().add(data);
        flushConfigFile(configurationsFile);
    }

    private void onConnectionEdit() {
		TableView.TableViewSelectionModel<ConnectionData> selectionModel = table.getSelectionModel();
        if (!selectionModel.isEmpty()) {
			int focusedIndex = selectionModel.getFocusedIndex();
            ConnectionData data = table.getItems().get(focusedIndex);
            NewConnectionBox connectionBox = (NewConnectionBox) beanFactory.getBean("newConnectionBox",
                    this, (ConnectionDataHandler) this::onConnectionEditData);
            connectionBox.update(data);
        }
    }

	private void onConnectionEditData(ConnectionData newData) {
        TableView.TableViewSelectionModel<ConnectionData> selectionModel = table.getSelectionModel();
        if (!selectionModel.isEmpty()) {
            int focusedIndex = selectionModel.getFocusedIndex();
            ConnectionData data = table.getItems().get(focusedIndex);
            data.setKeyspace(newData.getKeyspace());
            data.setPassword(newData.getPassword());
            data.setUrl(newData.getUrl());
            data.setUsername(newData.getUsername());
			// hack: better way to refresh the display??
			table.getColumns().get(0).setVisible(false);
			table.getColumns().get(0).setVisible(true);
		}           
        flushConfigFile(configurationsFile);
    }

    private void onConnectionUsage() {
        doWithSelected(connectionData -> {
            MainController controller = beanFactory.getBean(MainController.class);
            controller.loadTables(connectionData);
        });
    }

    private void onConnectionRemove() {
        doWithSelected(connectionData -> {
            table.getItems().remove(connectionData);
            flushConfigFile(configurationsFile);
        });
    }

    private void doWithSelected(Consumer<ConnectionData> callback) {
        TableView.TableViewSelectionModel<ConnectionData> selectionModel = table.getSelectionModel();

        if (!selectionModel.isEmpty()) {
            int focusedIndex = selectionModel.getFocusedIndex();
            ConnectionData connectionData = table.getItems().get(focusedIndex);
            callback.accept(connectionData);
        }
    }

    @SneakyThrows
    private void flushConfigFile(File configurationsFile) {
        Files.write(toJson(table.getItems()), configurationsFile, UTF_8);
    }

    @Override
    public void setBeanFactory(BeanFactory beanFactory) throws BeansException {
        this.beanFactory = beanFactory;
    }
}

File: src/main/kotlin/com/github/kindrat/cassandra/client/ui/window/menu/file/ConnectionManager.kt
package com.github.kindrat.cassandra.client.ui.window.menu.file

import com.datastax.oss.driver.api.core.type.codec.TypeCodecs
import com.github.kindrat.cassandra.client.i18n.MessageByLocaleService
import com.github.kindrat.cassandra.client.properties.StorageProperties
import com.github.kindrat.cassandra.client.properties.UIProperties
import com.github.kindrat.cassandra.client.ui.ConnectionData
import com.github.kindrat.cassandra.client.ui.MainController
import com.github.kindrat.cassandra.client.ui.fx.CellFactory.create
import com.github.kindrat.cassandra.client.ui.window.menu.ConnectionDataHandler
import com.github.kindrat.cassandra.client.ui.window.menu.KeySpaceProvider
import com.github.kindrat.cassandra.client.util.JsonUtil.fromJson
import com.github.kindrat.cassandra.client.util.JsonUtil.toJson
import javafx.beans.property.SimpleObjectProperty
import javafx.collections.FXCollections
import javafx.event.EventHandler
import javafx.geometry.Pos
import javafx.scene.Scene
import javafx.scene.control.*
import javafx.scene.image.Image
import javafx.scene.input.MouseEvent
import javafx.scene.layout.VBox
import javafx.stage.Modality
import javafx.stage.Stage
import mu.KotlinLogging
import org.apache.commons.lang3.StringUtils
import org.springframework.beans.BeansException
import org.springframework.beans.factory.BeanFactory
import org.springframework.beans.factory.BeanFactoryAware
import java.io.File
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.StandardOpenOption.WRITE
import java.util.function.Consumer
import java.util.function.Function

private val logger = KotlinLogging.logger {}

class ConnectionManager(
        parent: Stage,
        private val localeService: MessageByLocaleService,
        private val keySpaceProvider: KeySpaceProvider,
        private val uiProperties: UIProperties,
        private val storageProperties: StorageProperties
) : Stage(), BeanFactoryAware {

    private val table: TableView<ConnectionData>
    private val configurationsFile: File
    private var beanFactory: BeanFactory? = null

    @Throws(BeansException::class)
    override fun setBeanFactory(beanFactory: BeanFactory) {
        this.beanFactory = beanFactory
    }

    private fun getConfigurationsFile(): File {
        return File(storageProperties.location, storageProperties.propertiesFile)
    }

    private fun getButton(text: String): Button {
        val button = Button(text)
        button.maxWidth = uiProperties.connectionManagerButtonWidth
        button.minWidth = uiProperties.connectionManagerButtonWidth
        return button
    }

    private fun buildConnectionDataTable(leftWidth: Int): TableView<ConnectionData> {
        val dataTableView = TableView<ConnectionData>()
        dataTableView.minWidth = leftWidth.toDouble()
        dataTableView.maxWidth = leftWidth.toDouble()
        dataTableView.columnResizePolicy = TableView.UNCONSTRAINED_RESIZE_POLICY
        dataTableView.selectionModel.selectionMode = SelectionMode.SINGLE
        dataTableView.selectionModel.isCellSelectionEnabled = false
        dataTableView.onMouseClicked = EventHandler { mouseEvent: MouseEvent ->
            if (mouseEvent.clickCount > 1) {
                onConnectionUsage()
            }
        }
        val colWidth = leftWidth / 4
        val urlTitle = localeService.getMessage("ui.menu.file.manager.table.url")
        val url = buildColumn(urlTitle, colWidth) { data: ConnectionData -> data.url }
        val spaceTitle = localeService.getMessage("ui.menu.file.manager.table.keyspace")
        val keyspace = buildColumn(spaceTitle, colWidth) { data: ConnectionData -> data.keyspace }
        val dcTitle = localeService.getMessage("ui.menu.file.manager.table.dc")
        val dc = buildColumn(dcTitle, colWidth) { data: ConnectionData -> data.localDatacenter }
        val userTitle = localeService.getMessage("ui.menu.file.manager.table.username")
        val username = buildColumn(userTitle, colWidth) { data: ConnectionData -> data.username }
        val passTitle = localeService.getMessage("ui.menu.file.manager.table.password")
        val password = buildColumn(passTitle, colWidth) { data: ConnectionData -> data.password }
        dataTableView.columns.add(url)
        dataTableView.columns.add(keyspace)
        dataTableView.columns.add(dc)
        dataTableView.columns.add(username)
        dataTableView.columns.add(password)
        dataTableView.items = FXCollections.observableArrayList()
        return dataTableView
    }

    private fun buildColumn(name: String, columnWidth: Int, extractor: Function<ConnectionData, String?>): TableColumn<ConnectionData, String> {
        val column = TableColumn<ConnectionData, String>()
        val columnLabel = Label(name)
        column.cellFactory = create(TypeCodecs.TEXT)
        column.setCellValueFactory { param: TableColumn.CellDataFeatures<ConnectionData, String> ->
            val field = extractor.apply(param.value)
            SimpleObjectProperty(field)
        }
        column.setOnEditCommit { logger.info("edit") }
        column.graphic = columnLabel
        column.minWidth = columnWidth.toDouble()
        return column
    }

    private fun onNewConnection() {
        val handler = ConnectionDataHandler { handleConnectionData(it) }
        beanFactory!!.getBean("newConnectionBox", this, handler, keySpaceProvider)
    }

    private fun handleConnectionData(data: ConnectionData) {
        table.items.add(data)
        flushConfigFile(configurationsFile)
    }

    private fun onConnectionEdit() {
        val selectionModel = table.selectionModel
        if (!selectionModel.isEmpty) {
            val focusedIndex = selectionModel.focusedIndex
            val data = table.items[focusedIndex]
            val handler = ConnectionDataHandler { onConnectionEditData(it) }
            val connectionBox = beanFactory!!.getBean("newConnectionBox", this, handler, keySpaceProvider) as NewConnectionBox
            connectionBox.update(data)
        }
    }

    private fun onConnectionEditData(newData: ConnectionData) {
        val selectionModel = table.selectionModel
        if (!selectionModel.isEmpty) {
            val focusedIndex = selectionModel.focusedIndex
            val data = table.items[focusedIndex]
            val updatedData = data.copy(
                    url = newData.url,
                    keyspace = newData.keyspace,
                    localDatacenter = newData.localDatacenter,
                    username = newData.username,
                    password = newData.password
            )
            table.items[focusedIndex] = updatedData
            // hack: better way to refresh the display??
            table.columns[0].isVisible = false
            table.columns[0].isVisible = true
        }
        flushConfigFile(configurationsFile)
    }

    private fun onConnectionUsage() {
        doWithSelected { connectionData: ConnectionData ->
            val controller = beanFactory!!.getBean(MainController::class.java)
            controller.loadTables(connectionData)
            close()
        }
    }

    private fun onConnectionRemove() {
        doWithSelected { connectionData: ConnectionData ->
            table.items.remove(connectionData)
            flushConfigFile(configurationsFile)
        }
    }

    private fun doWithSelected(callback: Consumer<ConnectionData>) {
        val selectionModel = table.selectionModel
        if (!selectionModel.isEmpty) {
            val focusedIndex = selectionModel.focusedIndex
            val connectionData = table.items[focusedIndex]
            callback.accept(connectionData)
        }
    }

    private fun flushConfigFile(configurationsFile: File) {
        Files.write(configurationsFile.toPath(), toJson(table.items).toByteArray(StandardCharsets.UTF_8), WRITE)
    }

    init {
        initModality(Modality.APPLICATION_MODAL)
        initOwner(parent)
        title = localeService.getMessage("ui.menu.file.manager.title")
        icons.add(Image("cassandra_ico.png"))
        val width = uiProperties.connectionManagerWidth.toInt()
        val height = uiProperties.connectionManagerHeight.toInt()
        val rightWidth = uiProperties.connectionManagerRightPaneWidth.toInt()
        val leftWidth = width - rightWidth
        val connectionBox = VBox(uiProperties.newConnectSpacing)
        connectionBox.alignment = Pos.CENTER
        connectionBox.minWidth = leftWidth.toDouble()
        connectionBox.maxWidth = leftWidth.toDouble()
        table = buildConnectionDataTable(leftWidth)
        configurationsFile = getConfigurationsFile()
        if (!configurationsFile.exists()) {
            configurationsFile.createNewFile()
        }
        val json = Files.readString(configurationsFile.toPath())
        if (!StringUtils.isBlank(json)) {
            val connections = fromJson(json, ConnectionData::class.java)
            table.items.addAll(connections)
        }
        connectionBox.children.add(table)
        val buttons = VBox()
        buttons.minWidth = rightWidth.toDouble()
        buttons.maxWidth = rightWidth.toDouble()
        buttons.spacing = uiProperties.connectionManagerButtonsSpacing
        buttons.alignment = Pos.CENTER
        val newConnection = getButton("+")
        newConnection.onAction = EventHandler { onNewConnection() }
        val removeConnection = getButton("-")
        removeConnection.onAction = EventHandler { onConnectionRemove() }
        val editConnection = getButton("✎")
        editConnection.onAction = EventHandler { onConnectionEdit() }
        val useConnection = getButton("✓")
        useConnection.onAction = EventHandler { onConnectionUsage() }
        buttons.children.add(newConnection)
        buttons.children.add(removeConnection)
        buttons.children.add(editConnection)
        buttons.children.add(useConnection)
        val splitPane = SplitPane(connectionBox, buttons)
        val content = Scene(splitPane, width.toDouble(), height.toDouble())
        scene = content
        show()
    }
}